### ISSUE TITLE
Should there be additional timeout in case of LIBSSH2_ERROR_EAGAIN ?

### ISSUE THREAD

**Issue by rookierook** [libssh2/libssh2#544]
Hi Folks,

I am working with libssh2, I had some queries around the blocking calls, will really appreciate to get some answers:

How does the timeout mechanism for Non-Blocking API work, what error code is returned when the API times out on it's own?

Do you recommend additional timeout in case API is returning LIBSSH2_ERROR_EAGAIN, while we wait on socket on every LIBSSH2_ERROR_EAGAIN?

Thanks In Advance !

![carbon](https://user-images.githubusercontent.com/13089467/101279459-6d434e80-37e8-11eb-9b12-90aac8759e8d.png)

**Comment by pkittenis** [comment]
In non-blocking mode the client needs to handle timeouts in addition to what was set via `libssh2_session_set_timeout`.

Libssh2 will return `EAGAIN` on socket timeout in non-blocking mode *as well as* when the socket is blocked, if there is data to read in addition to the socket timeout. So there is no way to tell if an `EAGAIN` was because the socket timed out, or because it is otherwise blocked on reading/writing.

In other words, the client should implement timeout when handling `EAGAIN` which would be for the entire block of code shown above.

**Comment by willco007** [comment]
If you look at `_libssh2_wait_socket` it handles the timeout case with the correct error and is called from `BLOCK_ADJUST_ERRNO`. Which is called in the `*__ex` versions of the channel API.  If you use those APIs does it accomplish what you're looking for?
